keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md

```diff
@@ -0,0 +1,504 @@
+- [x] (R) Production readiness review completed
+- [x] (R) Production readiness review approved
+- [x] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+CronJob creates Jobs based on the schedule specified by the author, but the Time
+Zone used during the creation depens on where kube-controller-manager is running.
+This proposal aims to extend CronJob resource with the ability for a user to
+define the TimeZone when a Job should be created.
+
+## Motivation
+
+Not long after the [introduction of CronJob in kubernetes](https://github.com/kubernetes/kubernetes/pull/11980)
+a [request was raised to support setting time zones](https://github.com/kubernetes/kubernetes/issues/47202).
+The initial [response from SIG-Apps and SIG-Architecture](https://github.com/kubernetes/kubernetes/issues/47202#issuecomment-360820586)
+at the time was that introducing such functionality would require cluster operator
+to manually include TimeZone database since golang did not have one.
+Starting from [golang 1.15](https://go.dev/doc/go1.15) `time/tzdata` package can
+be embedded in the binary thus removing the requirement for external database.
+At that time the majority of the focus was towards [moving CronJob to GA](https://github.com/kubernetes/enhancements/issues/19)
+so the effort to support TimeZone was again delayed. Now that we have CronJob
+fully GA-ed it's time to satisfy the original request.
+
+### Goals
+
+- Add the field `.spec.timeZone` which allows specifying a valid TimeZone
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+Add the field `.spec.timeZone` to the CronJob resource. The cronjob controller
+will take the field into account when scheduling the next Job run. In case the
+field is not specified or is empty the controller will maitain the current
+behavior which is to rely on the time zone of the kube-controller-manager
+process.
+
+### Notes/Constraints/Caveats (Optional)
+
+The current mechanism, which will still be the default behavior heavily relies
+on the time zone of the kube-controller-manager process which is hard for a
+regular user to figure out. Exposing an explicit field for setting a time zone
+will allow CronJob authors to simplify the user experience when creating CronJobs.
+
```

Comment:
Would it make sense to also introduce a way to set a global default TZ?

Reply:
Ah, I had an idea about that.
_In my view_ that is best left for an external controller or for an admission webhook. However, the “external” mechanism could still be official Kubernetes code.

Reply:
Yes, no global setting, we don't have it anywhere, and I'd leave it to cluster admin to implement, re-use such setting.